### PR TITLE
HPCC-28014 Fix potential crashes if reset() called before stop

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
One change is to ensure that parent::stop() is called after all other actions - the reasoning is that all threads etc that may be reading from the input need to have stopped before stopping the input.   parent::reset() is called before any other actions - which should force the activity top stop if it hasn't been stopped already.
Fixes to potential crashes in keyedjoin (by moving the reset) and soapcall (by implementing stop())

Needs careful checking!

**Review by richardkchapman — APPROVED** [review]

**Review by mckellyln — APPROVED** [review]
Looks ok to me.
Just had one question.

**Review by afishbeck — APPROVED** [review]

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -12404,8 +12399,8 @@ class CRoxieServerIndexWriteActivity : public CRoxieServerInternalSinkActivity,
 
     virtual void reset()
     {
-        noteStatistic(StNumDuplicateKeys, cummulativeDuplicateKeyCount);
         CRoxieServerActivity::reset();
+        noteStatistic(StNumDuplicateKeys, cummulativeDuplicateKeyCount);
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp:12403]: Not new but...

Feels like cumuativeDuplicateKeyCount should be cleared here? Though an index write in a child query is unlikely...

Also that's not how you spell cumulative

Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:12403]: I agree.  I'll push a separate fix for that.

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -8135,16 +8135,11 @@ class CRoxieServerNormalizeChildActivity : public CRoxieServerActivity
         cursor = helper.queryIterator();
     }
 
-    virtual void stop()
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:8138]: correct - this implementation did nothing extra from the default.

**Pull request merged**